### ISSUE TITLE
Spike: using Angular JS to modify ExLibris views/pages

### ISSUE THREAD

**Issue by eporter23** [emory-libraries/blacklight-catalog#788]
The Alma Services page and Primo New UI both provide a customization package where CSS and JS can be added. We have been able to modify the styling of existing elements using CSS, but also may need to inject our own HTML elements into these views using JS. 

For this ticket, we want to investigate and confirm how to use and modify Angular JS as enabled within ExLibris applications. See the js/custom.js file in the package noted below.

As a proof of concept, try to achieve the following:

Add a div with link to the Alma Services Page header which goes back to the Blacklight homepage
`<div id=“bl-back-link”><a href=“http://blackcat-test.library.emory.edu”>Back to Catalog</a></div>`

Add a div with a link to the Alma Services page footer area that goes back to the Emory Libraries homepage
`<div id=“eul-footer”><a href=“https://libraries.emory.edu”>Emory Libraries</a></div>`

Emily, Lisa and Clare can assist with uploading the modified package files to Alma for testing. 

**Links and Examples**

- Emory Libraries GitHub - [package file repo](https://github.com/emory-libraries/alma-services-page) - contains a readme with some basic examples
- [ExLibris GitHub for customization package](https://github.com/ExLibrisGroup/primo-explore-package/blob/master/VIEW_CODE/js/README.md)
- [ExLibris iGelu presentation on working with angular](https://www.youtube.com/watch?v=dsRmHblEn_Y)
- [New Primo UI Cookbook](https://docs.google.com/document/d/1z1D5II6rhRd2Q01Uqpb_1v6OEFv_OksujEZ-htNJ0rw/edit#heading=h.xtcb9acz0ly2)
- [Other Primo New UI notes and tips](https://docs.google.com/document/d/1pfhN1LZSuV6ZOZ7REldKYH7TR1Cc4BUzTMdNHwH5Bkc/edit#)
- ExLibris Github - [Primo Explore Dev Environment ](https://github.com/ExLibrisGroup/primo-explore-devenv)
- [ExLibris - press release about Angular 1 end of life](https://knowledge.exlibrisgroup.com/Primo/Product_Materials/Announcements/Primo_with_AngularJS)
- Bernardo Gomez's Angular modifications on [Primo New UI test view](https://emory-primosb-alma.hosted.exlibrisgroup.com/primo-explore/search?vid=bernardo)
- [Example from Colorado state](example from https://libguides.colostate.edu/c.php?g=701450&p=6569946)

**Other Information**
- ExLibris using Angular v.1 and also incorporates some Google material design pattern markup (elements starting with <md-...>).
- Note: the ExLibris angular directives tend to follow the naming convention of <prm-element-name-before> or <prm-element-name-after> but all code examples show these names converted to camelCase such as prmElementNameBefore or prmElementNameAfter.
- Many of the ExLibris examples are targeting angular markup that doesn't exist in the Alma Services page view we are using (such as the search bar area).

**Comment by alexBLR** [comment]
If you want to play around with New Primo UI go to http://primo-studio.exlibrisgroup.com/
It is the easiest option to customize every single part (CSS, JS, HTML)

For developers, who want to run it in the local environment, you need to

1.  Clone https://github.com/ExLibrisGroup/primo-explore-devenv
2.  Follow installation instructions from that repository
3.  After installing the repo and its dependencies, please familiarize yourself with Git Submodules (https://www.atlassian.com/git/tutorials/git-submodule)
4. CD to primo-explore-devenv/custom
5. git submodule add https://github.com/emory-libraries/alma-services-page
6. You can now do your development, see the updated views and push the module back to our github when done
7. to publish the package after your development, run `gulp create-package `You will be prompted with a menu specifying all of the possible packages you can build
8. Log into Primo Back Office and navigate to the UI customization Package manager section : Primo Home > Primo Utilities > UI customization Package Manager and upload the new zipped package file.

**Comment by eporter23** [comment]
@alexBLR can you add a screenshot if possible that shows some modifications being added to a page? And also, to confirm, I am not able to modify JS in Primo Studio directly because the files show as read only - is that what you are seeing also?

**Comment by alexBLR** [comment]
New UI Github example for @eporter23: https://github.com/cbaksik/HVD2

Links customization: https://github.com/cbaksik/HVD2/blob/master/js/prm-service-links-after.js#L64-L70

Custom HTML: https://github.com/cbaksik/HVD2/blob/master/html/prm-service-links-after.html

**Comment by alexBLR** [comment]
Uploaded Harvard zipped repo https://github.com/cbaksik/HVD2 to primo-studio @eporter23 
![Screen Shot 2021-08-17 at 10.17.22 AM.png](https://images.zenhubusercontent.com/5c488275c10ae96cd915967b/6181f810-edac-4881-94e2-93d54eeee2cc)

**Comment by eporter23** [comment]
@alexBLR and anyone working on this, I spent some time looking at these examples and trying to apply them from the Alma-Summon based Services page. Unfortunately, these Primo-oriented options don't seem to work within the Alma views. I was, however, able to go back to Bernardo's custom.js file, pare it way down, and get a few things to load. In doing so I also found by trial and error that some of the prm- sections don't seem to be modifiable (like the actual footer), but others are. I was able to add some elements  near the logo and after the main content of the page however. Here's my [current working draft view](https://emory-psb.alma.exlibrisgroup.com/discovery/openurl?institution=01GALI_EMORY&vid=01GALI_EMORY:blacklightEP&rft.mms_id=9937294718502486).

All this said, I am sure the Primo-oriented options you've identified will be very helpful if we use the Primo VE Services Page.
